shopping-cart/shopping-cart-java/shopping-cart/src/main/java/com/example/shoppingcart/impl/ShoppingCart.java

```diff
@@ -0,0 +1,381 @@
+package com.example.shoppingcart.impl;
+
+import akka.actor.typed.ActorRef;
+import akka.cluster.sharding.typed.javadsl.EntityContext;
+import akka.cluster.sharding.typed.javadsl.EntityTypeKey;
+import akka.persistence.typed.PersistenceId;
+import akka.persistence.typed.javadsl.*;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.google.common.base.Preconditions;
+import com.lightbend.lagom.javadsl.persistence.AggregateEvent;
+import com.lightbend.lagom.javadsl.persistence.AggregateEventShards;
+import com.lightbend.lagom.javadsl.persistence.AggregateEventTag;
+import com.lightbend.lagom.javadsl.persistence.AggregateEventTagger;
+import com.lightbend.lagom.serialization.CompressedJsonable;
+import com.lightbend.lagom.serialization.Jsonable;
+import lombok.Value;
+import org.pcollections.HashTreePMap;
+import org.pcollections.PMap;
+
+import java.time.Instant;
+import java.util.Map;
+import java.util.Optional;
+
+public class ShoppingCart extends EventSourcedBehaviorWithEnforcedReplies<ShoppingCart.Command, ShoppingCart.Event, ShoppingCart.State> {
+
+    // We need to keep the original business id because that's the one we should
+    // use when tagging. If we use the PersistenceId, we will change the tag shard
+    private final String cartId;
+
+    private static final String ENTITY_TYPE_HINT = "ShoppingCartEntity";
+
+    // This is important for LagomJava, separator must be an empty String
+    private static final String PERSISTENCE_ID_SEPARATOR = "";
+
+    public static EntityTypeKey<Command> ENTITY_TYPE_KEY = EntityTypeKey.create(Command.class, ENTITY_TYPE_HINT);
+
+    public ShoppingCart(String cartId) {
+        this(cartId, PersistenceId.of(ENTITY_TYPE_HINT, cartId, PERSISTENCE_ID_SEPARATOR));
```

Comment:
```suggestion
        this(cartId, PersistenceId.of(ENTITY_TYPE_HINT, cartId));
```

Prefer using the default (see my comment above)